### PR TITLE
Document list of rules by what-they-do groups

### PR DESCRIPTION
**Pull request by jeddy3**
> Which issue, if any, is this issue related to?

Ref: https://github.com/stylelint/stylelint/issues/6376#issuecomment-1263543829

> Is there anything in the PR that needs further explanation?

This pull request groups our list of rules by what they do rather than what they're applied to.

I believe this approach helps our users more easily:

- understand what the built-in rules do
- find related built-in rules

You can [view the page here](https://github.com/stylelint/stylelint/blob/5111971b8b97c996a1792504b02d9e936df0b300/docs/user-guide/rules.md).

I think it does a better job of communicating the number of powerful rules we have for enforcing non-stylistic conventions that are unique to Stylelint.

This change also surfaced a lot of inconsistencies in our rule descriptions, which I've fixed in this pull request.

Grouping by the thing the rule applies to was mostly redundant as the rule name starts with it. I've still used those sub groups for the larger `no-*` and `*-list` group for readability, though. And I kept the old system for the stylistic rules as they'll be removed at some point.

The list of rules page now leads with just enough information for users to understand what they're looking at, i.e.:
- what makes a Stylelint rule
- and how it's named

I've moved other information that lived on the "About rules" page to the "Working on rules" page in the developer guide, which feel like a better home for it as it's more useful for people writing rules than using them.

### CODE REVIEW

File: docs/user-guide/rules.md [hunk 1]
```diff
@@ -0,0 +1,467 @@
...
+- [`custom-media-pattern`](../../lib/rules/custom-media-pattern/README.md): Specify a pattern for custom media query names.
+- [`custom-property-pattern`](../../lib/rules/custom-property-pattern/README.md): Specify a pattern for custom properties.
+- [`keyframes-name-pattern`](../../lib/rules/keyframes-name-pattern/README.md): Specify a pattern for keyframe names.
+- [`selector-class-pattern`](../../lib/rules/selector-class-pattern/README.md): Specify a pattern for class selectors.
+- [`selector-id-pattern`](../../lib/rules/selector-id-pattern/README.md): Specify a pattern for ID selectors.
+- [`selector-nested-pattern`](../../lib/rules/selector-nested-pattern/README.md): Specify a pattern for the selectors of rules nested within rules.
+
+### Quotes
+
+Require or disallow quotes with these `*-quotes` rules
```
Comment (ybiquitous) [inline docs/user-guide/rules.md:247]:
```suggestion
Require or disallow quotes with these `*-quotes` rules.
```

File: docs/user-guide/rules.md [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+- [`declaration-property-unit-allowed-list`](../../lib/rules/declaration-property-unit-allowed-list/README.md): Specify a list of allowed property and unit pairs within declarations.
+- [`declaration-property-unit-disallowed-list`](../../lib/rules/declaration-property-unit-disallowed-list/README.md): Specify a list of disallowed property and unit pairs within declarations.
+- [`declaration-property-value-allowed-list`](../../lib/rules/declaration-property-value-allowed-list/README.md): Specify a list of allowed property and value pairs within declarations.
+- [`declaration-property-value-disallowed-list`](../../lib/rules/declaration-property-value-disallowed-list/README.md): Specify a list of disallowed property and value pairs within declarations.
+
+#### Function
+
+- [`function-allowed-list`](../../lib/rules/function-allowed-list/README.md): Specify a list of allowed functions.
+- [`function-disallowed-list`](../../lib/rules/function-disallowed-list/README.md): Specify a list of disallowed functions.
+- [`function-url-no-scheme-relative`](../../lib/rules/function-url-no-scheme-relative/README.md): Disallow scheme-relative urls.
```
Comment (ybiquitous) [inline docs/user-guide/rules.md:151]: [note] Seeing the descriptions of `function-url-scheme-*`, using the word "URL" seems more consistent. Although this description change may be out of scope of this PR.

```suggestion
- [`function-url-no-scheme-relative`](../../lib/rules/function-url-no-scheme-relative/README.md): Disallow scheme-relative URLs.
```

File: docs/user-guide/rules.md [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+Disallow redundancy with these `no-redundant-*` rules.
+
+- [`declaration-block-no-redundant-longhand-properties`](../../lib/rules/declaration-block-no-redundant-longhand-properties/README.md): Disallow redundant longhand properties within declaration-block.
+- [`shorthand-property-no-redundant-values`](../../lib/rules/shorthand-property-no-redundant-values/README.md): Disallow redundant values within shorthand properties (Autofixable).
+
+## Enforce stylistic conventions
+
+We have frozen these rules — we won't fix bugs nor add options, and we will deprecate then remove them in future releases. We recommend you use a pretty printer (like Prettier) alongside Stylelint rather than these rules. If you prefer to use Stylelint to enforce stylistic consistency, you can [migrate the rules you need to a plugin](../developer-guide/plugins.md).
+
+The whitespace rules allow you to enforce an empty line, a single space, a newline or no space in some specific part of the stylesheet.
```
Comment (ybiquitous) [inline docs/user-guide/rules.md:264]: [suggest] The beginning of this sentence about the whitespace rules seemed slightly too abrupt. How about adding a sub-section header like "Whitespace" or something?

```suggestion
### Whitespace

The whitespace rules allow you to enforce an empty line, a single space, a newline or no space in some specific part of the stylesheet.
```

File: lib/rules/font-family-name-quotes/README.md [hunk 4]
```diff
@@ -1,6 +1,6 @@
 # font-family-name-quotes
 
-Specify whether or not quotation marks should be used around font family names.
+Require or disallow quotes for font family names
```
Comment (ybiquitous) [inline lib/rules/font-family-name-quotes/README.md:3]:
```suggestion
Require or disallow quotes for font family names.
```

File: lib/rules/font-family-no-duplicate-names/README.md [hunk 5]
```diff
@@ -1,6 +1,6 @@
 # font-family-no-duplicate-names
 
-Disallow duplicate font family names.
+Disallow duplicate names within font families
```
Comment (ybiquitous) [inline lib/rules/font-family-no-duplicate-names/README.md:3]:
```suggestion
Disallow duplicate names within font families.
```

File: lib/rules/font-family-no-missing-generic-family-keyword/README.md [hunk 6]
```diff
@@ -1,6 +1,6 @@
 # font-family-no-missing-generic-family-keyword
 
-Disallow missing generic families in lists of font family names.
+Disallow missing generic family keyword within font families.
```
Comment (ybiquitous) [inline lib/rules/font-family-no-missing-generic-family-keyword/README.md:3]: [suggest] Clearer by inserting `a`?

```suggestion
Disallow missing generic a family keyword within font families.
```

File: docs/user-guide/rules.md [hunk 7]
```diff
@@ -0,0 +1,467 @@
...
+### Missing
+
+Disallow missing things with these `*-no-missing` rules.
+
+- [`custom-property-no-missing-var-function`](../../lib/rules/custom-property-no-missing-var-function/README.md): Disallow missing `var` function for custom properties.
+- [`font-family-no-missing-generic-family-keyword`](../../lib/rules/font-family-no-missing-generic-family-keyword/README.md): Disallow missing generic family keyword within font families.
+
+### Nonstandard
+
+Disallow nonstandard things with these `*-no-nonstandard` rules.
```
Comment (ybiquitous) [inline docs/user-guide/rules.md:89]: [suggest] The word "non-standard" seems a little more readable, although the rule name is unchangeable.

```suggestion
### Non-standard

Disallow non-standard things with these `*-no-nonstandard` rules.
```

**Review by ybiquitous — COMMENTED** [review]
@jeddy3 Thank you for the significant improvements! 👏🏼 

I've left comments to suggest fixing trivial mistakes, but any comments don't block this PR. Please feel free to address or reject them.

**Comment by ybiquitous** [comment]
Note: Perhaps, we need to update `meta.url` in all rules later like this if rule URLs are changed in `stylelint.io`:

https://github.com/stylelint/stylelint/blob/659972918323c6a15015a16188d130d7f7700612/lib/rules/alpha-value-notation/index.js#L21-L22

```diff
- url: 'https://stylelint.io/user-guide/rules/list/alpha-value-notation',
+ url: 'https://stylelint.io/user-guide/rules/alpha-value-notation',
```

But then, because redirecting also will be added, it seems we can do it after publishing the new documents.

**Comment by jeddy3** [comment]
@ybiquitous Thanks for all the suggestions.

[Latest version](https://github.com/stylelint/stylelint/blob/f11fb5f1d8a66eb9083d7169e62d5af7061223de/docs/user-guide/rules.md).

> Note: Perhaps, we may need to update meta.url in all rules later 

Done also. The URL is nicer without `list`.

Prepped the stylelint.io repo: https://github.com/stylelint/stylelint.io/pull/285

File: lib/rules/shorthand-property-no-redundant-values/README.md [hunk 8]
```diff
@@ -1,6 +1,6 @@
 # shorthand-property-no-redundant-values
 
-Disallow redundant values in shorthand properties.
+Disallow redundant values within shorthand properties
```
Comment (ybiquitous) [inline lib/rules/shorthand-property-no-redundant-values/README.md:3]: Oops.

```suggestion
Disallow redundant values within shorthand properties.
```

**Review by ybiquitous — APPROVED** [review]
Thank you! LGTM 👍🏼

**Pull request merged**